docs/docs/higher-order-components.md

```diff
@@ -0,0 +1,421 @@
+
+Mutating HOCs are a leaky abstraction—the consumer must know how they are implemented in order to avoid conflicts with other HOCs.
+
+Instead of mutation, HOCs should use composition, by wrapping the input component in a container component:
+
+```js
+function logProps(WrappedComponent) {
+  return class extends React.Component {
+    componentWillReceiveProps(nextProps) {
+      console.log('Current props: ', this.props);
+      console.log('Next props: ', nextProps);
+    }
+    render() {
+      // Wraps the input component in a container, without mutating it. Good!
+      return <WrappedComponent {...this.props} />;
+    }
+  }
+}
+```
+
+This HOC has the same functionality as the mutating version while avoiding the potential for clashes. It works equally well with class and functional components. And because it's a pure function, it's composable with other HOCs, or even with itself.
+
+You may have noticed similarities between HOCs and a pattern called **container components**. Container components are part of a strategy of separating responsibility between high-level and low-level concerns. Containers manage things like subscriptions and state, and pass props to components that handle things like rendering UI. HOCs use containers as part of their implementation. You can think of HOCs as parameterized container component definitions.
+
+## Convention: Pass Unrelated Props Through to the Wrapped Component
+
+HOCs add features to a component. They shouldn't drastically alter its contract. It's expected that the component returned from an HOC has a similar interface to the wrapped component.
+
+HOCs should pass through props that are unrelated to its specific concern. Most HOCs contain a render method that looks something like this:
+
+```js
+render() {
+  // Filter out extra props that are specific to this HOC and shouldn't be
+  // passed through
+  const { extraProp, ...passThroughProps } = this.props;
+
+  // Inject props into the wrapped component. These are usually state values or
+  // instance methods.
+  const injectedProp = someStateOrInstanceMethod;
+
+  // Pass props to wrapped component
+  return (
+    <WrappedComponent
+      injectedProp={injectedProp}
+      {...passThroughProps}
+    />
+  );
+}
+```
+
+This convention helps ensure that HOCs are as flexible and reusable as possible.
+
+## Convention: Maximizing Composability
+
+Not all HOCs look the same. Sometimes they accept only a single argument, the wrapped component:
+
+```js
+const NavbarWithRouter = withRouter(Navbar);
+```
+
+Usually, HOCs accept additional arguments. In this example from Relay, a config object is used to specify a component's data dependencies:
+
+```js
+const CommentWithRelay = Relay.createContainer(Comment, config);
+```
+
+By far, the most common signature for HOCs looks like this:
```

Comment:
Is this actually the most common signature? I found this sentence confusing. How about just:

"Some HOCs, like React Redux's `connect`, look different:"